line_profiler/_line_profiler.pyx

```diff
@@ -156,72 +203,75 @@ cdef class LineProfiler:
 
     def enable(self):
         PyEval_SetTrace(python_trace_callback, self)
-
-    def disable(self):
-        self.last_time = {}
+        
+    @property
+    def code_map(self):
+        return <dict>self.c_code_map
+        
+    @property
+    def last_time(self):
+        return <dict>self.c_last_time
+
+    cpdef disable(self):
+        self.c_last_time.clear()
         unset_trace()
 
-    def get_stats(self):
+    cpdef get_stats(self):
         """ Return a LineStats object containing the timings.
         """
+        cdef dict cmap
+        
         stats = {}
-        for code in self.code_map:
-            entries = self.code_map[code].values()
+        for code in self.code_hash_map:
+            cmap = self.c_code_map
+            entries = []
+            for entry in self.code_hash_map[code]:
+                entries += list(cmap[entry].values())
             key = label(code)
-            stats[key] = [e.astuple() for e in entries]
+            stats[key] = [(e["lineno"], e["nhits"], e["total_time"]) for e in entries]
             stats[key].sort()
         return LineStats(stats, self.timer_unit)
 
-
-cdef class LastTime:
-    """ Record the last callback call for a given line.
-    """
-    cdef int f_lineno
-    cdef PY_LONG_LONG time
-
-    def __cinit__(self, int f_lineno, PY_LONG_LONG time):
-        self.f_lineno = f_lineno
-        self.time = time
-
-
+@cython.boundscheck(False)
+@cython.wraparound(False)
 cdef int python_trace_callback(object self_, PyFrameObject *py_frame, int what,
-    PyObject *arg):
+PyObject *arg):
     """ The PyEval_SetTrace() callback.
     """
     cdef LineProfiler self
-    cdef object code, key
-    cdef dict line_entries, last_time
-    cdef LineTiming entry
+    cdef object code
+    cdef LineTime entry
     cdef LastTime old
+    cdef int key
     cdef PY_LONG_LONG time
+    cdef int64 code_hash
+    cdef int64 block_hash
+    cdef unordered_map[int64, LineTime] line_entries
 
     self = <LineProfiler>self_
-    last_time = self.last_time
 
     if what == PyTrace_LINE or what == PyTrace_RETURN:
-        code = <object>py_frame.f_code
-        if code in self.code_map:
+        block_hash = hash(get_frame_code(py_frame))
+        code_hash = block_hash ^ py_frame.f_lineno
```

Comment:
Honestly, I think any bitwise operator would work. The only requirement is that code_hash needs to not be the same as any other line in a program. Essentially, `block_hash` should already be unique for essentially any python code block, and XORing it with the line number is a quick way to combine the block hash and the line number. I admit I don't have a mathematical or pseudo-cryptographic basis for any of my arguments, so if you know anybody who could perform an analysis of this code, I'd greatly appreciate it.

Reply:
Also, let's consider that normally we don't have more than like 10000 lines being profiled at a single time (I rarely profile more than 500 or so). We have 2^64 potential hashes from the block_hash alone, so a back-of-the-envelope calculation would *probably* yield a very low chance of collision.

Reply:
This makes sense to me now. From a practical perspective I think you're right that we could probably ship this as-is and never run into an issue... but the topic piqued my interest, and I'm not sure if this theoretically weakens the hash at all.

If I wanted to be very safe, I would hash a tuple of the code and line number. That would guarantee the strength of the hash. But there is a performance penalty there. We probably don't want to construct tuples in the `python_trace_callback` loop.

A look at https://stackoverflow.com/questions/5889238/why-is-xor-the-default-way-to-combine-hashes provides some interesting insight.

From a theoretical perspective, the block hash should a fairly random "good hash". However, the line number - which we are effectively treating as a hash - won't be. According to the [hash_guidelines](https://www.javamex.com/tutorials/collections/hash_function_guidelines.shtml), we should spread the line number bits out a bit by multiplying them with some constant. 

If you feel fancy, I think we can afford a few more shifts and additions without adding too much overhead, so maybe we can do something like the "boost::hash_combine" mentioned in that post?

(But I should also go to bed... I actually tried, but I kept thinking about this question :upside_down_face: )

Reply:
I'll try benchmarking the boost::hash_combine solution, thanks for finding that SO post! My guess is it'll be acceptable overhead, and I believe any downside could be negated by switching the unordered_map counts to short-circuiting finds (which I was already planning to do).

Reply:
So boost::hash_combine when copied to cython generates an overflow error, because iirc the Python hash can be up to 2^64-1, and that overflows one of the bit shifts. I would try out the other suggestion, 3*hash(a) + hash(b), however I noticed this quote:
> So hash(a) + hash(b) is better than hash(a) xor hash(b) in that if a==b, the result is hash(a)<<1 instead of 0.

We know that the block hash is an extremely large number, somewhere around 2^63, and the line number is a number that is almost always less than ~10000. The case where a == b is simply not going to happen.

Reply:
It still might be good to multiply by a prime number like 53. My thought is that doing otherwise probably does weaken the hash by a small amount, and taking this small step would recover most of it. Granted, it probably wouldn't weaken it by much, but I'd be you can prove you'd get about 9 less bits of entropy for ints in the range 10,000, and after multiplying I bet that number would be a lot closer to 0.

Reply:
Yeah, that's reasonable, let me test that.